---
title: "22_08___RNAseq_T1RS_ADIctrl"
output: html_notebook
---

Raw data fasta files -> aligned with STAR -> sorted, indexed and filtered to unqiuely mapped reads in samtools

```{r}
#library(Rsubread)
library(R.utils)
library(tidyverse)
library(ggfortify)
library(ggrepel)
library(DESeq2)
library(ashr)
library(EnhancedVolcano)
library(org.Hs.eg.db)
library(OrganismDbi)
library(clusterProfiler)
library(pathview)


```

Make references to bam files and gtf file

```{r}
cwd <- getwd()
#gtf_file <- gzip(filename = file.path(cwd, "annotation_gtf/Homo_sapiens.GRCh38.104.gtf.gz"),
#                 destname = file.path(cwd, "annotation_gtf/unzipped_GRCH38.gtf"),
#                 remove = FALSE,
#                 overwrite = TRUE)
gtf_file <- paste0(cwd, "/annotation_gtf/unzipped_GRCH38.gtf")
bam_files <- list.files(path = paste0(cwd, "/data/sorted_unique_mapped_bam_files"), full.names = TRUE)

```

Run featureCounts from Rsubread package - launching R on the HPC (below is a R script to be used with the Rscript command on unix), not that BiocManager needs to be installed on unix first

```{r}
install.packages("BiocManager")
BiocManager::install("Rsubread")
library(Rsubread)
gtf_file <- "/data/BCI-WrenchLab/GRCh38_ensembl/Homo_sapiens.GRCh38.104.gtf"
bam_files <- list.files(path = "/data/BCI-WrenchLab/maustin/jun21_PDX_rnaseq_sorted_unique_bam_files/", full.names = TRUE)
fc <- featureCounts(bam_files,
                    annot.inbuilt = "hg38",
                    annot.ext = gtf_file[1],
                    isGTFAnnotationFile = TRUE,
                    GTF.featureType = "exon",
                    GTF.attrType = "gene_id",
                    useMetaFeatures = TRUE, #summarise at gene level
                    isPairedEnd = TRUE)
save(fc, file = "/data/BCI-WrenchLab/maustin/jobs/Rsubread/Rsubread_output.Rdata")

```


Next job is to load the data generated by Rsubread on the HPC into R

```{r}
cwd <- getwd()
load(file.path(cwd, "Rsubread_output_jul22.Rdata"))

```

The columns in the count data matrix are headed by the unique IDs sent to the sequencing company. These need to be renamed.

```{r}

countData <- fc$counts
sampleIDs <- c("A3", "A4", "A5", "A6", "A7", "A10",
              "A8", "A9", "A11", "A12", "A1", "A2")

colnames(countData) <- sampleIDs
```

Need a metaData file for what these samples represent

```{r}
library(tidyverse)

cellLineID <- c("TOM-1","TOM-1","TOM-1","TOM-1","RS4;11","RS4;11","RS4;11","RS4;11","RS4;11","RS4;11","TOM-1","TOM-1")
treatmentInfo <- c("Control","ADI","ADI","ADI","Control","ADI","Control","Control","ADI","ADI","Control","Control")
  
sampleInfo <- tibble(sampleIDs, cellLineID, treatmentInfo)

```

Some QC on reads

```{r}
#BiocManager::install("DESeq2")
#library(DESeq2)

#remove genes with low expression, just for QC images, DESeq2 deals with this itself later
keep <- rowSums(countData) >5
countData_lowStrippped <- countData[keep,]

#simple plot of raw counts to look for outliers
bp <- boxplot(countData_lowStrippped, main='Raw counts distribution', las = 2)

#raw counts means vs SD
plot(rowMeans(countData_lowStrippped),
     rowSds(countData_lowStrippped),
     main = 'Raw counts: Mean vs SD', xlim = c(0,10000), ylim = c(0,5000))

#qc plots of rlog transformed data
rlog_counts <- rlog(countData_lowStrippped)
statusCol <- match(sampleInfo$treatmentInfo, c("ADI","Control")) +1
boxplot(rlog_counts,
        main = 'Rlog counts',
        xlab = "",
        ylab = "rlog counts",
        las = 2,
        col = statusCol)
abline(h=median(rlog_counts))

plot(rowMeans(rlog_counts),
     rowSds(rlog_counts),
     main = 'rlog counts: Mean vs SD',
     xlab = 'Mean',
     ylab = 'SD',
     xlim = c(0,20), ylim = c(0,6))


```

A basic PCA plot to see if the controls worked and the effect of the biological replicates

```{r}
#library(ggfortify)
#library(ggrepel)

pcDat <- prcomp(t(rlog_counts))
PCA_plot <- autoplot(pcDat,
         data = sampleInfo,
         colour = "cellLineID",
         shape = "treatmentInfo",
         size = 5) +
  geom_text_repel(aes(x=PC1, y=PC2, label=sampleIDs), box.padding = 0.8, max.overlaps = 15)


PCA_plot

```

Differential gene expression analysis. We add an extra group variable to the sampleInfo tibble so we can easily extract the ADI vs control changes per PDX

```{r}
sampleInfo$group <- c("TOM1.Ctrl","TOM1.ADI","TOM1.ADI","TOM1.ADI","RS411.Ctrl","RS411.ADI",
                      "RS411.Ctrl","RS411.Ctrl","RS411.ADI","RS411.ADI","TOM1.Ctrl","TOM1.Ctrl")

design <- as.formula(~ group)
modelMatrix <- model.matrix(design, data = sampleInfo)

ddsObj.raw <- DESeqDataSetFromMatrix(countData = countData,
                                     colData = sampleInfo,
                                     design = design)
ddsObj <- DESeq(ddsObj.raw)

res_TOM1_AvC <- results(ddsObj,
                         contrast = c("group","TOM1.ADI","TOM1.Ctrl"))

shrunken_res_TOM1_AvC <- lfcShrink(ddsObj,
                         contrast = c("group","TOM1.ADI","TOM1.Ctrl"),
                         type = "ashr")

topGenes_TOM1_AvC <- as.data.frame(res_TOM1_AvC) %>%
  rownames_to_column("GeneID") %>%
  arrange(padj) %>%
  drop_na()

topGenes_TOM1_AvC$symbol <- AnnotationDbi::mapIds(org.Hs.eg.db,
                                         keys = topGenes_TOM1_AvC$GeneID,
                                         column = "SYMBOL",
                                         keytype = "ENSEMBL",
                                         multiVals = "first")

res_RS411_AvC <- results(ddsObj,
                         contrast = c("group","RS411.ADI","RS411.Ctrl"))

shrunken_res_RS411_AvC <- lfcShrink(ddsObj,
                         contrast = c("group","RS411.ADI","RS411.Ctrl"),
                         type = "ashr")

topGenes_RS411_AvC <- as.data.frame(res_RS411_AvC) %>%
  rownames_to_column("GeneID") %>%
  arrange(padj) %>%
  drop_na()

```

Plot some results

```{r}
# plot the expression of ASS1
plotCounts(ddsObj, gene = "ENSG00000130707", intgroup = "group",)

# MA plots on shrukne LFCs to visualise data
plotMA(shrunken_res_TOM1_AvC, ylim = c(-2,2))
plotMA(shrunken_res_RS411_AvC, ylim = c(-0.1,0.1))


```

Look at the upregulated and downregulated genes in each case

```{r}
TOM1_up <- topGenes_TOM1_AvC[topGenes_TOM1_AvC$log2FoldChange > 1,]
TOM1_up <- TOM1_up[TOM1_up$padj < 0.05,]
TOM1_up$symbol <- AnnotationDbi::mapIds(org.Hs.eg.db,
                                         keys = TOM1_up$GeneID,
                                         column = "SYMBOL",
                                         keytype = "ENSEMBL",
                                         multiVals = "first")
TOM1_up$entrez <- AnnotationDbi::mapIds(org.Hs.eg.db,
                                         keys = TOM1_up$GeneID,
                                         column = "ENTREZID",
                                         keytype = "ENSEMBL",
                                         multiVals = "first")

# 

TOM1_down <- topGenes_TOM1_AvC[topGenes_TOM1_AvC$log2FoldChange < -1,]
TOM1_down <- TOM1_down[TOM1_down$padj < 0.05,]
TOM1_down$symbol <- AnnotationDbi::mapIds(org.Hs.eg.db,
                                         keys = TOM1_down$GeneID,
                                         column = "SYMBOL",
                                         keytype = "ENSEMBL",
                                         multiVals = "first")
TOM1_down$entrez <- AnnotationDbi::mapIds(org.Hs.eg.db,
                                         keys = TOM1_down$GeneID,
                                         column = "ENTREZID",
                                         keytype = "ENSEMBL",
                                         multiVals = "first")

# T8017 has 108 downregulated genes

#------------------------------------------

RS411_up <- topGenes_RS411_AvC[topGenes_RS411_AvC$log2FoldChange > 0,]
RS411_up <- RS411_up[RS411_up$padj < 0.05,]
#RS411_up$symbol <- AnnotationDbi::mapIds(org.Hs.eg.db,
#                                         keys = RS411_up$GeneID,
#                                         column = "SYMBOL",
#                                         keytype = "ENSEMBL",
#                                         multiVals = "first")
#RS411_up$entrez <- AnnotationDbi::mapIds(org.Hs.eg.db,
#                                         keys = RS411_up$GeneID,
#                                         column = "ENTREZID",
#                                         keytype = "ENSEMBL",
#                                         multiVals = "first")



#RS411_down <- topGenes_RS411_AvC[topGenes_RS411_AvC$log2FoldChange < 0,]
#RS411_down <- RS411_down[RS411_down$padj < 0.05,]
#RS411_down$symbol <- AnnotationDbi::mapIds(org.Hs.eg.db,
#                                         keys = RS411_down$GeneID,
#                                         column = "SYMBOL",
#                                         keytype = "ENSEMBL",
#                                         multiVals = "first")
#RS411_down$entrez <- AnnotationDbi::mapIds(org.Hs.eg.db,
#                                         keys = RS411_down$GeneID,
#                                         column = "ENTREZID",
#                                         keytype = "ENSEMBL",
#                                         multiVals = "first")



# Make an euler diagram



```

Volcano plots

```{r}


labels_of_interest <- c("TNFRSF10B","CEBPB","MAP1LC3B","PLAU","ASNS",
                        "TRIB3","CHAC1","SIRT1","ATF3")
lab_italics <- paste0("italic('", topGenes_TOM1_AvC$symbol, "')")
selectLab_italics = paste0(
    "italic('",
    labels_of_interest,
    "')")

EV <- EnhancedVolcano(topGenes_TOM1_AvC,
                lab = topGenes_TOM1_AvC$symbol,
                x = 'log2FoldChange',
                y = 'padj',
                pCutoff = 0.05,
                FCcutoff = 1,
                legendPosition = '',
                selectLab = labels_of_interest,
                boxedLabels = TRUE,
                #parseLabels = TRUE,
                caption = '',
                drawConnectors = TRUE,
                labSize = 5,
                labFace = 'bold',
                col = c('grey', 'grey','grey', 'red3'),
                title = "",
                subtitle = "Labels: ATF4 target genes, padj = 3.6e-6"
                ) + ylim(0,75)

```

Do Kegg analysis in clusterProfiler

```{r}
#sigGenes_TOM1_up <- TOM1_up %>%
#  drop_na(entrez, padj) %>%
#  filter(abs(log2FoldChange) >1 & padj <0.01) %>%
#  arrange(log2FoldChange) #%>%
#  pull(entrez)
#kegg_TOM1_up <- enrichKEGG(gene = sigGenes_TOM1_up)

#sigGenes_TOM1_down <- TOM1_down %>%
#  drop_na(entrez, padj) %>%
#  filter(abs(log2FoldChange) >1 & padj <0.01) %>%
#  arrange(log2FoldChange) #%>%
#  pull(entrez)
#kegg_TOM1_down <- enrichKEGG(gene = sigGenes_TOM1_down)

sig_genes_tom1 <- topGenes_TOM1_AvC %>%
  drop_na(entrez, padj) %>%
  filter(padj <0.001, log2FoldChange >(1)) %>%
  arrange(log2FoldChange) %>%
  pull(entrez)
kegg_tom1_avc <- enrichKEGG(gene = sig_genes_tom1, organism = 'hsa')
d <- dotplot(kegg_tom1_avc, showCategory = 20)

tom1_gsea_ora <- enricher(sig_genes_tom1, TERM2GENE = H_t2g)
d_gsea_ora <- dotplot(tom1_gsea_ora, showCategory = 20)

#---------------------------------------------------

#sigGenes_C3442_up <- C3442_up %>%
#  drop_na(entrez, padj) %>%
#  filter(abs(log2FoldChange) >0.5 & padj <0.05) %>%
#  arrange(log2FoldChange) %>%
#  pull(entrez)
#kegg_C3442_up <- enrichKEGG(gene = sigGenes_C3442_up)

#sigGenes_C3442_down <- C3442_down %>%
#  drop_na(entrez, padj) %>%
#  filter(abs(log2FoldChange) >0.5 & padj <0.05) %>%
#  arrange(log2FoldChange) %>%
#  pull(entrez)
#kegg_C3442_down <- enrichKEGG(gene = sigGenes_C3442_down)

#--------------------------------

#T8017_up_FC_entrez <- T8017_up$log2FoldChange
#names(T8017_up_FC_entrez) <- T8017_up$entrez
#pathview(gene.data = T8017_up_FC_entrez,
#         pathway.id = "hsa04110",
#         limit = list(gene=2,cpd=1),
#         species = "hsa")

```

GSEA data prep for genepattern website

```{r}

gsea_T8017 <- topGenes_T8017_AvC[topGenes_T8017_AvC$padj <0.1,] %>%
  arrange(desc(log2FoldChange))

# add symbols
gsea_T8017$symbol <- AnnotationDbi::mapIds(org.Hs.eg.db,
                                         keys = gsea_T8017$GeneID,
                                         column = "SYMBOL",
                                         keytype = "ENSEMBL",
                                         multiVals = "first")


gsea_C3442 <- topGenes_C3442_AvC[topGenes_C3442_AvC$padj <0.1,] %>%
  arrange(desc(log2FoldChange))

# add symbols
gsea_C3442$symbol <- AnnotationDbi::mapIds(org.Hs.eg.db,
                                         keys = gsea_C3442$GeneID,
                                         column = "SYMBOL",
                                         keytype = "ENSEMBL",
                                         multiVals = "first")


```

```{r}
# GSEA in clusterprofiler

library(msigdbr)
H_t2g <- msigdbr(species = "Homo sapiens", category = "H") %>% 
  dplyr::select(gs_name, entrez_gene)

#topGenes_TOM1_AvC$entrez <- AnnotationDbi::mapIds(org.Hs.eg.db,
#                                         keys = topGenes_TOM1_AvC$GeneID,
#                                         column = "ENTREZID",
#                                         keytype = "ENSEMBL",
#                                         multiVals = "first")
gsea_genes_tom1 <- topGenes_TOM1_AvC %>%
  drop_na(entrez, padj) %>%
  filter(padj <0.05) %>%
  arrange(log2FoldChange)
gsea_list <- gsea_genes_tom1$log2FoldChange
names(gsea_list) <- gsea_genes_tom1$entrez
gsea_list <- gsea_list[unique(names(gsea_list))]
gsea_list <- sort(gsea_list, decreasing = TRUE)
tom1_gsea <- GSEA(gsea_list, TERM2GENE = H_t2g)
d_gsea <- dotplot(tom1_gsea, showCategory = 20)

hub_genes_list <- topGenes_TOM1_AvC %>%
  drop_na(entrez, padj) %>%
  filter(padj < 0.05) %>%
  filter(abs(log2FoldChange) >1) %>%
  arrange(log2FoldChange)
gsea_list <- hub_genes_list$log2FoldChange
names(gsea_list) <- hub_genes_list$entrez
gsea_list <- gsea_list[unique(names(gsea_list))]
gsea_list <- sort(gsea_list, decreasing = TRUE)
tom1_hub_gsea <- GSEA(gsea_list, TERM2GENE = H_t2g)
d_gsea <- dotplot(tom1_hub_gsea, showCategory = 20)

```

```{r}

require(DOSE)
d <- dotplot(tom1_gsea, showCategory=50, split=".sign") + facet_grid(.~.sign)


```


```{r}
#Extract genes from specific pathways

E2F_t2g <- filter(H_t2g, gs_name == "HALLMARK_E2F_TARGETS")
TNFa_t2g <- filter(H_t2g, gs_name == "HALLMARK_TNFA_SIGNALING_VIA_NFKB")

TNFa_full <- topGenes_TOM1_AvC[topGenes_TOM1_AvC$entrez %in% TNFa_t2g$entrez_gene,]
TNFa_full$symbol <- AnnotationDbi::mapIds(org.Hs.eg.db,
                                         keys = TNFa_full$GeneID,
                                         column = "SYMBOL",
                                         keytype = "ENSEMBL",
                                         multiVals = "first")
TNFa_full <- TNFa_full[order(TNFa_full$log2FoldChange, decreasing = TRUE),]

TNFa_trunc <- select(TNFa_full, symbol, log2FoldChange, padj)



E2F_full <- topGenes_TOM1_AvC[topGenes_TOM1_AvC$entrez %in% E2F_t2g$entrez_gene,]
E2F_full$symbol <- AnnotationDbi::mapIds(org.Hs.eg.db,
                                         keys = E2F_full$GeneID,
                                         column = "SYMBOL",
                                         keytype = "ENSEMBL",
                                         multiVals = "first")
E2F_full <- E2F_full[order(E2F_full$log2FoldChange, decreasing = TRUE),]

E2F_trunc <- select(E2F_full, symbol, log2FoldChange, padj)


```
